### ISSUE TITLE
Much faster borg backup over SSH via ProxyCommand's use of nc

### ISSUE THREAD

**Issue by ttsiodras** [borgbackup/borg#3747]
This is not a bug - it's more likely a pull request in disguise :-)

I am using borgbackup to do a weekly backup of my Arch Linux laptop contents to an Orange PI SBC running Armbian. I am doing this over borgbackup, and recently realized that the SSH speeds used are far inferior to what my wifi link can deliver.  You can [read my debugging story here](https://unix.stackexchange.com/questions/434825/ssh-speed-greatly-improved-via-proxycommand-but-why), but the executive summary is that I got almost an order of magnitude faster transfers via use of netcat in SSH's ProxyCommand:

    $ cat .ssh/config
    Host orangepi
        Hostname 192.168.1.150
        User root
        Port 22
        ProxyCommand nc %h %p

....which can also be done in the command line - as my [debugging story shows](https://unix.stackexchange.com/questions/434825/ssh-speed-greatly-improved-via-proxycommand-but-why) via `ssh -o "ProxyCommand nc %h %p"`.

Perhaps you can have a look and see if this improves speed in your tests as well - it definitely did on my own.

**Comment by dragetd** [comment]
For reference: The Orange Pi Zero has a quadcore Cortex-A7 which is moderately quick. I used to use a Raspberry Pi 1 which struggled badly with SSH and Borg to get over 600kb/s.

Since Benchmarks show that the Orage Pi Zero is not too far off from the Pine64, which I am currently using as a backup target, I would assume it would handle 2-5mb/S which I am currently pushing to mine.

In your Stack Exchange post you assume it might be the driver detecting SSH traffic and by going through NC it is somehow packaged differently… ? This is a rather scary and strange theory. SSH-traffic is encrypted in software and there is not much to be done with it on the network layer… but that being said, the most crazy stupid theory would be the Chinese actually messing with SSH in their binary blob kernel. Are there sources for the same kernel available? If we'd manage to compile it from source and verify the same effect, that we could investigate the source.

Adding nc ProxyCommand to my setup did not change much at all, tho I am limited by my upload capacity.

I doubt that this is a borg-specific effect.

**Comment by ttsiodras** [comment]
Hi, @dragetd - no, I just finished debugging this further.

Look at my post again - I just updated the answer, fully tracking this down. It was SSH in the end - it was disabling Nagle's algorithm...

Have a look at the final form of my answer - I honestly think that it in the end it does apply to all users of SSH as a transport that deal with slow links (e.g. slow/frail WiFi).

**Comment by klausenbusk** [comment]
> I am unsure how to do it "formally", though... I wish SSH had a command line option to do this properly - (i..e. --dont-disable-nagle). But in its absence, I think the closest "hack" is my use of nc under ProxyCommand.

At least some version of SSH contain a `NoDelay` option (https://docstore.mik.ua/orelly/networking_2ndEd/ssh/ch07_04.htm). I think you should open a feature request upstream: https://bugzilla.mindrot.org/

**Comment by ttsiodras** [comment]
@klausenbusk  Done: 

https://bugzilla.mindrot.org/show_bug.cgi?id=2848

**Comment by ThomasWaldmann** [comment]
This sounds interesting. Guess we need some more data for this from other users - post your link types and the speed you get with borgbackup using the 2 different ways to setup the ssh connection.

**Comment by ttsiodras** [comment]
Makes sense, Thomas - perhaps it merits a mention in the docs even now. 

But in relation to your comment about SBCs, I believe I've proven conclusively in my debugging that use of an SBC at the remote end had nothing to do with the phenomenon. Look at the [output of dstat](https://i.stack.imgur.com/4CaQt.png) - during the test the SBC reported less than 25% CPU utilization of only one of its 4 cores.

Put simply, SSH is currently always using `setsockopt(...., TCP_NODELAY)`, no matter what the link is. In slow WiFis, this means that the link is immediately transmitting whatever packet data are sent to the socket - and depending on the quality of the link, that can end up wasting a significant part of the available bandwidth (_think of wifi packets that are mostly headers_).

I'll just leave [this asciicast here as the best possible proof](https://asciinema.org/a/rMpxMHNXIBAaOiBvSjAISbLBU) I can provide.

**Comment by stevesbrain** [comment]
@ThomasWaldmann Sorry, are you after tests with + without disabled Nagle, or with + without the netcat ProxyCommand?

**Comment by ThomasWaldmann** [comment]
@stevesbrain as far as i understood, using netcat was the only way to enable it. if it could be done within ssh, that would be of course the more direct way.

**Comment by ttsiodras** [comment]
@stevesbrain You can indeed use the LD_PRELOAD "hack" that I described in my post - but I doubt we'd want this monstrosity to become a part of borgbackup :-) 

So indeed, Thomas is right - try using the ProxyCommand. If the WiFi link you are using is fast enough, you many not see much of a  difference - but if it's a lousy one, you should. The lousier the WiFi link, the more Nagle should help with bandwidth...

**Comment by ThomasWaldmann** [comment]
The docs which is read by most people is the one for the latest release, so we have some time left until next release.

**Comment by stevesbrain** [comment]
I'll see if I can get some useful stats :)

**Comment by jdchristensen** [comment]
In my tests, over wifi, I get the same speed with ordinary ssh and with the nc proxy.

I'm skeptical that this will affect many people.  If you are feeding data to ssh fast enough, then it shouldn't matter whether you use Nagle or not.  And ssh has been so extensively used that if this was a common problem, it would be well known.  I think there must be something fishy going on in @ttsiodras 's setup, maybe with the driver for that particular wifi chip.  The fact that packet counts weren't very different with and without Nagle suggest that that might be a red herring, and it might instead be a bug that's triggered only when Nagle is off.

**Comment by ThomasWaldmann** [comment]
Also interesting would be a independent reproduction of the problem on similar hardware / software as @ttsiodras used (esp. considering that it was not very special).

**Comment by ttsiodras** [comment]
@jdchristensen Well, I have a bootable USB stick with and older version of SystemRescue, so here goes - I just booted with it, configured my WiFi settings, connected, and [repeated the exercise](    https://drive.google.com/file/d/1tT0s4pz9hIQ6MPcYnVEE5uMsqfsCyJiX/view)

At the very least you now know this is not "fake" :-) - and that this happens with different distros and different kernels (I believe the SystemRescue stick I am using has a 4.9 kernel).

**Comment by ThomasWaldmann** [comment]
Considering that this was not reproducible in general and there was no feedback of other affected users / users who could gain a significant speedup by using this, I am closing this now.
